velox/exec/Unnest.h

```diff
@@ -39,22 +39,70 @@ class Unnest : public Operator {
   bool isFinished() override;
 
  private:
-  // Generate output for 'size' input rows starting from 'start' input row.
+  // Represents the range of rows to process and indicates that first and last
+  // row may need to be processed partially to match the output batch size.
+  // The row range is firstRowStart_ to lasRowEnd when there is
+  // single row to process, moreover, the row range for first row is
+  // firstRowStart_ to rawMaxSizes_[firstRow] and the row range for last row is
+  // 0 to lastRowEnd when there are several rows to process unless the last row
+  // process fully, in that case, use rawMaxSizes_[lastRow] as end of the last
+  // row.
   //
-  // @param start First input row to include in the output.
-  // @param size Number of input rows to include in the output.
-  // @param outputSize Pre-computed number of output rows.
-  RowVectorPtr generateOutput(
-      vector_size_t start,
-      vector_size_t size,
-      vector_size_t outputSize);
+  // Single row:
+  //    firstRowStart_  firstRowEnd = lastRowEnd
+  //---|----------------|--- start, size = 1
+  //
+  // Multiple rows:
+  //    firstRowStart_     firstRowEnd = rawMaxSizes_[start]
+  //---|-------------------| start
+  //-----------------------
+  //-----------------------
+  //-----------------|----- end - 1
+  //                 lastRowEnd
+  // The size is end - start, the range is [start, end).
+  struct RowRange {
+    // Invokes a function on each row represent by the RowRange.
+    // @param func Call the function for each row, `row` means the current row
+    // number whose range is [start, start + size), `start` means the row number
+    // to start to process, `size` means the number of input rows to process.
+    // @param rawMaxSizes Used to compute the end of each row.
+    // @param firstRowStart Same with Unnest member firstRowStart_, start
+    // processing the first row input from `firstRowStart_`.
+    void forEachRow(
+        std::function<void(
+            vector_size_t /*row*/,
+            vector_size_t /*start*/,
+            vector_size_t /*size*/)> func,
+        const vector_size_t* const rawMaxSizes,
+        vector_size_t firstRowStart) const;
+
+    // First input row to be included in the output.
+    const vector_size_t start;
+
+    // Number of input rows to be included in the output.
+    const vector_size_t size;
+
+    // Processing of the last input row begins at index firstRowStart_ or 0
+    // depending on whether the row to process is first row and ends at
+    // 'lastRowEnd'. It is nullopt when processing the last row fully.
```

Comment:
> // The processing of the last input row starts at index 'firstRowStart_' or 0,
// depending on whether it is the first row being processed, and ends at 'lastRowEnd'.
// It is nullopt when the last row is processed completely.